### PR TITLE
feat: add onError option

### PR DESCRIPTION
**Pull request by kangmingtay**
* Adds an `onError` option which accepts a callback function for error handling purposes 
* This will be very useful for handling retries on the client 
* This is a non-breaking change since `onError` is an optional parameter and is only invoked if it's defined

### CODE REVIEW

File: src/http_client.ts [hunk 1]
```diff
@@ -34,6 +38,12 @@ class LogflareHttpClient {
     protected readonly apiKey: string
     protected readonly fromBrowser: boolean
     protected readonly apiBaseUrl: string
+    /**
+     * onError takes in an optional callback function to handle any errors returned by logflare
+     */
+    protected readonly onError?: ((payload: {
+        batch: object[];
+    }, err: Error) => void) | undefined
```
Comment (Ziinc) [inline src/http_client.ts:46]: typing looks off, should be `null`

File: src/http_client.ts [hunk 2]
```diff
@@ -91,14 +102,19 @@ class LogflareHttpClient {
             return data
         } catch (e) {
             if (e) {
```
Comment (Ziinc) [inline src/http_client.ts:104]: we can inverse the logic to return early so that there is less `if` levels

File: src/http_client.ts [hunk 3]
```diff
@@ -91,14 +102,19 @@ class LogflareHttpClient {
...
+                    if (e instanceof NetworkError && e.response) {
+                        console.error(
+                            `Logflare API request failed with ${
+                                e.response.status
+                            } status: ${JSON.stringify(e.data)}`
+                        )
+                    } else {
+                        console.error(e.message)
+                    }
+                    if (this.onError) {
```
Comment (Ziinc) [inline src/http_client.ts:115]: i think we can simplify this to a one liner `this.onError?.(payload, e)`

**Review by Ziinc — CHANGES_REQUESTED** [review]

File: src/http_client.ts (hunk 1 above)
Comment (kangmingtay) [reply src/http_client.ts:45]: oops, intended to use `undefined` instead of `null` since this option would default to `undefined` if nothing was passed in

Comment (kangmingtay) [reply src/http_client.ts:45]: i've removed `onError` from the `defaultOptions` since it's not used

File: src/http_client.ts (hunk 2 above)
Comment (Ziinc) [reply src/http_client.ts:103]: not quite sure what you mean
```ts
// no error -> return
// not an Error instance -> return
if (!e && !(e instance of Error)) return;
// remaining error checking
if (e instance of Network Error && e.response) {
  ...
} else {
  ...
}
// fire callback
this.onError?.(...)
```

**Review by Ziinc — COMMENTED** [review]
could we update the tests as well to cover the callback firing

**Comment by kangmingtay** [comment]
@Ziinc haha sure idm doing that, just feels harder to read with the double negatives

**Comment by Ziinc** [comment]
gotcha. That is fine then.

**Review by Ziinc — APPROVED** [review]

**Pull request merged**